### DISCUSSION TITLE
Get the `stargazerCount`s for multiple repositories

### DISCUSSION THREAD

**Discussion by abiriadev** [community/community#68283] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

## What I'm trying to do

I'm working on a toy project to extend the current GitHub code search API by adding some sorting options.

For instance, I'd like to sort the results based on the stargazer count of the repositories. Using the REST API, I would have to send individual requests for each result, which probably lead to a rate limit issue. So I've decided to use GraphQL.

This is the GraphQL query I have come up with so far:

```graphql
query MultileStars {
  rest_api_description: repository(owner: "github", name: "rest-api-description") {
    stargazerCount
  }
  vscode: repository(owner: "microsoft", name: "vscode") {
    stargazerCount
  }
}
```

The result is:
```json
{
  "data": {
    "rest_api_description": {
      "stargazerCount": 1217
    },
    "vscode": {
      "stargazerCount": 151467
    }
  }
}
```

While it works, I can't be sure this is the best way to fetch multiple results at once. The more repositories I want to fetch, the longer the query becomes.

Is there a better way to achieve this? Any suggestions or opinions would be greatly appreciated.

Regards.

**Comment by LangLangBart** [comment]
- When you have received a response from the Github code search using the `REST API`, you can create an array with all the `node_ids` and then loop over that array to create a single `Graphql` query.

```graphql
{
  node1: node(id: "R_kgDOArmXAg") {
    ... on Repository {
      nameWithOwner
      stargazerCount
    }
  }
}
```

- for example, search for files called `config.py`, and display up to  10 results

```sh
#!/usr/bin/env bash

# Create a temporary file
temp_graphql_query_file=$(mktemp)

# Use trap command to ensure the temporary file is removed when script exits
trap 'rm -f "$temp_graphql_query_file"' EXIT

# Start the query
echo "query {" >"$temp_graphql_query_file"

index=1
gh api search/code --method GET \
	--raw-field q='filename:config.py' \
	--header "X-Github-Next-Global-ID:1" \
	--field per_page=10 \
	--jq '.items[].repository.node_id' |
	while read -r node_id; do
		{
			echo "  Node${index}: node(id: \"$node_id\") {"
			echo "    ... on Repository {"
			echo "      nameWithOwner"
			echo "      stargazerCount"
			echo "    }"
			echo "  }"
		} >>"$temp_graphql_query_file"
		index=$((index + 1))
	done
# End the query
echo "}" >>"$temp_graphql_query_file"

# Call the generated graphql file
gh api graphql --raw-field query="$(cat "$temp_graphql_query_file")"

# Alternative command, with '@' to upload data from a file, same result as the command above
# gh api graphql --field query=@"$temp_graphql_query_file"
```

```json
  {
    "data": {
      "Node1": {
        "nameWithOwner": "tensorflow/tensorflow",
        "stargazerCount": 177798
      },
      "Node2": {
        "nameWithOwner": "scikit-learn/scikit-learn",
        "stargazerCount": 55937
      },
      "Node3": {
        "nameWithOwner": "pytorch/pytorch",
        "stargazerCount": 71071
      },
      "Node4": {
        "nameWithOwner": "v8/v8",
        "stargazerCount": 21720
      },
      "Node5": {
        "nameWithOwner": "Significant-Gravitas/AutoGPT",
        "stargazerCount": 149360
      },
      "Node6": {
        "nameWithOwner": "home-assistant/core",
        "stargazerCount": 63061
      },
      "Node7": {
        "nameWithOwner": "emscripten-core/emscripten",
        "stargazerCount": 24272
      },
      "Node8": {
        "nameWithOwner": "SFTtech/openage",
        "stargazerCount": 11983
      },
      "Node9": {
        "nameWithOwner": "3b1b/manim",
        "stargazerCount": 53861
      },
      "Node10": {
        "nameWithOwner": "commaai/openpilot",
        "stargazerCount": 41864
      }
    }
  }
```

---

- above is too complex, a better solution would be if GitHub adds the `CODE` search ability to the `GraphQL API` 😃 

```sh
gh api graphql --raw-field query=$'{__type(name: "SearchType") { enumValues { name } }}' \
	--jq '[.data.__type.enumValues[].name] | join(", ")'
# ISSUE, REPOSITORY, USER, DISCUSSION
```

**Reply by abiriadev** [reply]
First of all, I appreciate your thoughtful and detailed answer.

However, this is exactly the approach I've taken. I'm wondering if this is truly the 'best way' to do this task because generating a graphql query feels similar to 'generating an SQL query'(I know this isn't a correct comparison, but it's how I feel.). If graphql has a built-in method for some kind of this 'iteration', I would prefer to use that.

May I ask you if this is the most acceptable approach you think? Since I am quite new to the GitHub graphql API(although I've used GitHub rest API several times) and haven't checked out all the available queries yet.

> above is too complex, a better solution would be if GitHub adds the `CODE` search ability to the `GraphQL API` 😃

I hope so too. :)

Regards.

**Comment by LangLangBart** [comment]
> If graphql has a built-in method for some kind of this 'iteration', I would prefer to use that.

- no, I don't see a 'better way `spec page` - https://spec.graphql.org/

>  Since I am quite new to the GitHub graphql API

Overall, the `GraphQL API` is great. One specific request and you get what you ask for. However, it has some shortcomings, such as the `CODE` search, lack of Notifications (#13056), and lack of support for Actions (#24493). These issues force you to juggle between two APIs, as their features overlap like a Venn Diagram.

---

Informative talk about GraphQL by Rea Loretta (15/Dec/19) - [Advanced patterns for GitHub's GraphQL API](https://www.youtube.com/watch?v=i5pIszu9MeM)

**Reply by abiriadev** [reply]
I got it. Thank you for taking your time!

**Comment by LangLangBart** [comment, marked as answer]
> no, I don't see a 'better way spec page - https://spec.graphql.org/

@abiriadev, I need to retract that statement. There is a simpler way by using a list type modifier for type checking and `nodes`.

<img src="https://github.com/community/community/assets/92653266/c671b787-ba91-46ea-9499-a9d205367ab3" width="400">

- Ref: [devhints.io/graphql](https://devhints.io/graphql)
- [Queries - GitHub Docs - Nodes](https://docs.github.com/en/graphql/reference/queries#nodes)

### Query 
```graphql
query ($ids: [ID!]!) {
  nodes(ids: $ids) {
    ... on Repository {
      nameWithOwner
      stargazerCount
    }
  }
}
```

#### Input Variables as a LIST
- these **node_ids** were collected from the `REST API` call and concatenated into a **LIST**

```json
{"ids": ["R_kgDOArmXAg", "R_kgDOAAzd1g", "R_kgDOA-j9zw", "R_kgDOAXSgmg", "R_kgDOAMSroQ", "R_kgDOAfLOxw", "R_kgDOJKSTjA", "R_kgDOAAGmTg", "R_kgDOJPJM2Q", "R_kgDOAA1TcA", "R_kgDOAS1I7A", "R_kgDOABS35A", "R_kgDOAA3dPw", "R_kgDOBEmZvQ", "R_kgDOAALQjA", "R_kgDOBtbnWg", "R_kgDOAAoMVg", "R_kgDODixnZQ", "R_kgDOBITxeA", "R_kgDOD0zRAg"]}
```

### Result
```json
{
  "data": {
    "nodes": [
      {
        "nameWithOwner": "tensorflow/tensorflow",
        "stargazerCount": 178438
      },
      {
        "nameWithOwner": "scikit-learn/scikit-learn",
        "stargazerCount": 56243
      },
      {
        "nameWithOwner": "pytorch/pytorch",
        "stargazerCount": 71998
      },
      {
        "nameWithOwner": "v8/v8",
        "stargazerCount": 21848
      },
      {
        "nameWithOwner": "home-assistant/core",
        "stargazerCount": 63571
      },
      {
        "nameWithOwner": "3b1b/manim",
        "stargazerCount": 54391
      },
      {
        "nameWithOwner": "Significant-Gravitas/AutoGPT",
        "stargazerCount": 151818
      },
      {
        "nameWithOwner": "mongodb/mongo",
        "stargazerCount": 24572
      },
      {
        "nameWithOwner": "twitter/the-algorithm",
        "stargazerCount": 60038
      },
      {
        "nameWithOwner": "getsentry/sentry",
        "stargazerCount": 35418
      },
      {
        "nameWithOwner": "odoo/odoo",
        "stargazerCount": 31174
      },
      {
        "nameWithOwner": "emscripten-core/emscripten",
        "stargazerCount": 24393
      },
      {
        "nameWithOwner": "numpy/numpy",
        "stargazerCount": 24822
      },
      {
        "nameWithOwner": "ray-project/ray",
        "stargazerCount": 28345
      },
      {
        "nameWithOwner": "celery/celery",
        "stargazerCount": 22420
      },
      {
        "nameWithOwner": "deepfakes/faceswap",
        "stargazerCount": 47340
      },
      {
        "nameWithOwner": "ipython/ipython",
        "stargazerCount": 15977
      },
      {
        "nameWithOwner": "mingrammer/diagrams",
        "stargazerCount": 31843
      },
      {
        "nameWithOwner": "llvm/llvm-project",
        "stargazerCount": 22465
      },
      {
        "nameWithOwner": "kiwibrowser/src",
        "stargazerCount": 2437
      }
    ]
  }
}
```

**Reply by LangLangBart** [reply]
This makes querying for multiple IDs much simpler, but adding the `CODE` search ability to the `GraphQL API` is still much preferred.

**Reply by abiriadev** [reply]
Thank you so much! I tested it out a few minutes ago, and it works the way exactly I wanted!
